### PR TITLE
UCP/RNDV: Option to limit GPUDirectRDMA size in rndv

### PR DESCRIPTION
**Pull request by bureddy**
## What
Option to avoid GPUDirectRDMA on large messages rndv

## How ?
or mem type, Ignore mem type GDR lane while initializing the lanes for zcopy if the length of the zcopy is greater than the threshold,

### CODE REVIEW

File: src/ucp/rndv/rndv.c [hunk 1]
```diff
@@ -20,6 +20,32 @@
...
 #include <ucs/datastruct/queue.h>
 
+static UCS_F_ALWAYS_INLINE int
+ucp_rndv_memtype_direct_support(ucp_context_h context, size_t reg_length,
+                                ucs_memory_type_t mem_type,
+                                uint64_t reg_mem_types)
+{
+    return ((mem_type == UCS_MEMORY_TYPE_HOST) ||
+            !(reg_mem_types & UCS_BIT(UCS_MEMORY_TYPE_HOST)) ||
+            (reg_length < context->config.ext.rndv_memtype_direct_size));
```
Comment (yosefe) [inline src/ucp/rndv/rndv.c:30]: why need so many checks (which add cpu overhead), can't just check `reg_length < context->config.ext.rndv_memtype_direct_size` and `(mem_type == UCS_MEMORY_TYPE_HOST)` in ucp_rndv_is_get_zcopy?

Comment (bureddy) [reply src/ucp/rndv/rndv.c:30]: we wantt another check do only for GDR lane and not on cuda-ipc lane

Comment (yosefe) [reply src/ucp/rndv/rndv.c:30]: how is GDR lane differentiated from cuda-ipc lane?

Comment (bureddy) [reply src/ucp/rndv/rndv.c:30]: cuda-ipc lane can't register HOST memory. assuming that only GDR lane can do HOST and MEMTYPE registration

Comment (yosefe) [reply src/ucp/rndv/rndv.c:30]: 1. it's very non-obvious (and not documented). can we check by sys_dev instead? for example if they have same sysdev than it's ok to use zcopy
2. in general need to reduce the overhead of this check: use bitmap operations, precompute into ep_config whatever possible, etc

**Comment by bureddy** [comment]
@yosefe can you please review

File: src/ucp/rndv/rndv.c (hunk 1 above)
Comment (bureddy) [reply src/ucp/rndv/rndv.c:33]: @yosefe As per our discussion, I added some verbose and mem type check in the function

File: src/ucp/rndv/rndv.c [hunk 2]
```diff
@@ -20,6 +20,39 @@
...
+}
+
+static UCS_F_ALWAYS_INLINE void
+ucp_rndv_memtype_direct_update_md_map(ucp_context_h context,
+                                      ucp_request_t *sreq, ucp_md_map_t *md_map)
+{
+    unsigned md_index;
+    uct_md_attr_t *md_attr;
+
+    if(UCP_MEM_IS_HOST(sreq->send.mem_type)) {
```
Comment (yosefe) [inline src/ucp/rndv/rndv.c:43]: space after if

File: src/ucp/rndv/rndv.c [hunk 3]
```diff
@@ -20,6 +20,39 @@
 #include <ucp/proto/proto_am.inl>
 #include <ucs/datastruct/queue.h>
 
+static UCS_F_ALWAYS_INLINE int
```
Comment (yosefe) [inline src/ucp/rndv/rndv.c:23]: space line before

File: src/ucp/rndv/rndv.c [hunk 4]
```diff
@@ -20,6 +20,39 @@
...
+ucp_rndv_memtype_direct_support(ucp_context_h context, size_t reg_length,
+                                ucs_memory_type_t mem_type,
+                                uint64_t reg_mem_types)
+{
+    /* for large mem type rndv, allow transfers on
+    ** mem type only support rndv lane, and
+    ** mem type direct lane if transfer length is less than threshold.
+    */
+    return (!(reg_mem_types & UCS_BIT(UCS_MEMORY_TYPE_HOST)) ||
```
Comment (yosefe) [inline src/ucp/rndv/rndv.c:32]: extra ( )

File: src/ucp/rndv/rndv.c [hunk 5]
```diff
@@ -20,6 +20,39 @@
...
 #include <ucs/datastruct/queue.h>
 
+static UCS_F_ALWAYS_INLINE int
+ucp_rndv_memtype_direct_support(ucp_context_h context, size_t reg_length,
+                                ucs_memory_type_t mem_type,
+                                uint64_t reg_mem_types)
+{
+    /* for large mem type rndv, allow transfers on
+    ** mem type only support rndv lane, and
+    ** mem type direct lane if transfer length is less than threshold.
```
Comment (yosefe) [inline src/ucp/rndv/rndv.c:30]: comment is hard to parse

File: src/ucp/rndv/rndv.c [hunk 6]
```diff
@@ -21,6 +21,40 @@
 #include <ucs/datastruct/queue.h>
 
 
+static UCS_F_ALWAYS_INLINE int
+ucp_rndv_memtype_direct_support(ucp_context_h context, size_t reg_length,
+                                ucs_memory_type_t mem_type,
```
Comment (yosefe) [inline src/ucp/rndv/rndv.c:26]: this param is not used

File: src/ucp/rndv/rndv.c [hunk 7]
```diff
@@ -21,6 +21,40 @@
...
+                                uint64_t reg_mem_types)
+{
+    /* for large mem type rndv, allow transfers if lane is
+    ** mem type only rndv lane which can't do host zcopy, or
+    ** length is less than threshold.
+    */
```
Comment (yosefe) [inline src/ucp/rndv/rndv.c:32]: How about:
If the message size is rndv_memtype_direct_size or larger, disable zero-copy rndv on lanes that also support host memory.

**Review by yosefe — APPROVED** [review]

**Review by yosefe — APPROVED** [review]
